### ISSUE TITLE
Rigging more than one function

### ISSUE THREAD

**Issue by krlmlr** [moodymudskipper/boomer#16]
that call each other.

https://github.com/moodymudskipper/boomer/issues/11#issuecomment-826821204

I'd like to propose a `wire()` function:

```r
wire <- function(..., ignore = , ) {
  names <- c(...)

  # rigs all functions
  # ensures that the function environments are updated with the newly rigged functions
}
```

And also `ignite()`, to be called from `.onLoad()`:

```r
ignite <- function() {
  config <- Sys.getenv("R_BOOMER")
  ...
  wire(...)  
}
```

To be used like this:

```
.onLoad <- function(...) {
  if (requireNamespace("boomer", quietly = TRUE)) {
    boomer::ignite()
  }
}
```

**Comment by moodymudskipper** [comment]
`wire` rigs in place and thus can take several inputs ?
I could see how it could be handy, I could trace the functions to return the output of a call to a rigged copy, and we'd use `untrace` to "unwire".

I don't understand what `ignite()` does, does it rig all the package's functions ?

The use case here is package development ?

I think for `wire`, if I understood correctly what it should be, I'd use a more boring name such as `trace_rig` or `rig_in_place` to make the difference between functions more obvious.

**Comment by krlmlr** [comment]
Yes.

`ignite()` takes inspiration from how {debugme} handles this: you can specify in an environment variable which functions to rig, and don't need to change the code to enable or disable rigging. This allows rigging an installed package too.

I wouldn't couple this too tightly with `trace()`. Also, since this is for package development, there's no need to support un-rig -- a fresh R session without the environment variable does the job.

**Comment by krlmlr** [comment]
How about `multi_rig()` or `rig_some()` ?

**Comment by moodymudskipper** [comment]
Thanks, I'll take a look at {debug me}, sounds interesting

**Comment by moodymudskipper** [comment]
If we  modify functions in their namespace : 
* We're doing weird things that CRAN might disapprove of
* [we won't be able to rig base functions](https://developer.r-project.org/blosxom.cgi/R-devel/NEWS/2021/04/16#n2021-04-16)
* We miss flexibility for unrigging some or all.

If we use devtools, I believe we can simply place rigged copies in an environment attached on the search path right next to `"devtools_shims"`, let's call it `"boomer_shims"`.

* it is less weird
* we can rig all functions
* building helpers to unrig, or have multi_rig work like `dplyr::select` by doing `multi_rig(-this_fun, -that_fun)`
This way we can also rig functions from other packages, including base.

The rigged function's body's functions should be evaluated first in `"boomer_shims"` and then in the rigged function's environment, so it can work with functions calling each other, I think we can use :

```
shim_list <- as.list(as.environment("boomer_shims"))
ns <- parent.environment(environment())
eval(fun_sym, shim_list , ns)
```
Rigging namespaced functions would be harder, e.g. if i want to `multi_rig(dplyr::filter)`.

To support it we might :
* have `multirig(dplyr::filter)`  add a function called `"dplyr::filter"` to `"boomer_shims"` (non syntactic, highly unlikely to conflict in practice, and easy to understand)
* override `::` and `:::` in `boomer_shims` so that they look for a match in `"boomer_shims"` before checking the actual namespaces.

It's still not quite enough because if I've attached dplyr or do it later, `filter` should find a shimmed copy in `"boomer_shims"` ,
so `multirig(dplyr::filter)`  should also create a copy of `filter`, but this copy should check if {dplyr} is attached and do the appropriate action (rigged dplyr::filter or simple stats::filter, explicit failure for cases where nothing on the search path).

`multi_rig` doesn't communicate that we do something specific to development, maybe something like `rig_shims` makes sense

**Comment by krlmlr** [comment]
This starts getting very complicated. The current implementation works great. I suspect shims might get us into other kinds of trouble down the line. I remember this painfully from mockr. Also, I don't care too much about unrigging.

I see two components of the problem:

1. Functions `a()` and `b()` a rigged, if `a()` calls `b()` it uses the original version of `b()` (or the other way round). I think you confirmed that?
2. A way to conveniently rig functions in your own package without changing the code.

The first can be achieved with a simple extension of `rig()` AFAICT: get all functions to be modified, rig them, update the environments for all functions, update them. Assigning to the own namespace during loading shouldn't bother CRAN too much.

For the second problem, {debugme} can provide inspiration.

**Comment by moodymudskipper** [comment]
@krlmlr I've implemented `rig_in_namespace`

I've played around a bit and it seems to work fine, 

**it's a bit hard to follow the log though**, maybe the rigged function should announce itself : we'd print in a different color "going through `foo(...)`", "going through `bar(...)`", "back to `foo(...)`". It would change the body though, so wouldn't be robust in some corner cases such as displaying a function's body.

Another option is to change colors each time we enter a different rigged function, 3 cycling colors would be enough for use to know if we're going forward or back in the stack.

**unexported functions when not using devtools::load_all()** (which makes basically exports them), is not well supported yet.

**I'd like a new `rig_in_namespace` call to cancel the previous one**, OR to add to it, depending on a parameter. For this we need to be able to unrig, but it'll be possible through https://github.com/moodymudskipper/boomer/issues/23

I think `rig_in_namespace` could work by default a bit like `dplyr::select`, we can use `-` to unrig, we could use tidy selection and we'd be able to rig what `starts_with()` or `ends_with()` something for instance (or use `everything()`). an argument `.keep = FALSE` triggers whether we start from already rigged functions or from scratch.

**Comment by krlmlr** [comment]
To unrig, can we store the old function code?

`rig_in_namespace()` doesn't yet work for me in `.onLoad()`. Should it?

**Comment by moodymudskipper** [comment]
> To unrig, can we store the old function code?

The code doesn't change, the environment does, and we'd need to store it indeed

**Comment by krlmlr** [comment]
Most functions use the package environment, but I agree it's safer to cache it for the general case.

**Comment by krlmlr** [comment]
Currently, it seems difficult to create a reprex with `rig_in_namespace()` . What can we do about that?

**Comment by moodymudskipper** [comment]
Maybe some wrappers around this could work, this seems reprex proof : 

```
# define `makeNamespace` function
eval(body(loadNamespace)[[c(8, 4, 4)]])

ns <- makeNamespace("fake")

# define functions
ns$foo <- function() "foo!"
ns$bar <- function() "bar!"

# export some
namespaceExport(ns, "foo")
#> <environment: 0x0000000011af0248>

# `:::` works
fake:::foo()
#> [1] "foo!"
fake:::bar()
#> [1] "bar!"

# `::` works on exported functions
fake::foo()
#> [1] "foo!"
fake::bar()
#> Error: 'bar' is not an exported object from 'namespace:fake'
```

**Comment by moodymudskipper** [comment]
done in 9010b2151f5281f634ad94286a04bab4792f5fbf

In the above I exported both functions, it's not a limitation of `fake_package()` but of `rig_in_namespace()`, for now it is not able to rig unexported functions (except when using devtools::load_all(), which basically exports all).

**Comment by krlmlr** [comment]
Could `rig_in_namespace()` also work for plain environments such as `.GlobalEnv` ?

**Comment by moodymudskipper** [comment]
The main feature is there, splitting the rest into new issues for V 0.2
